File: mobile/library/swift/src/EnvoyClient.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// Envoy's implementation of `StreamClient`, buildable using `StreamClientBuilder`.
@objcMembers
final class EnvoyClient: NSObject {
  private let engine: EnvoyEngine

  private enum ConfigurationType {
    case yaml(String)
    case typed(EnvoyConfiguration)
  }

  private init(configType: ConfigurationType, logLevel: LogLevel, engine: EnvoyEngine) {
    self.engine = engine
    super.init()

    switch configType {
    case .yaml(let configYAML):
      self.engine.run(withConfigYAML: configYAML, logLevel: logLevel.stringValue)
    case .typed(let config):
      self.engine.run(withConfig: config, logLevel: logLevel.stringValue)
    }
  }

  /// Initialize a new Envoy instance using a typed configuration.
  ///
  /// - parameter config:   Configuration to use for starting Envoy.
  /// - parameter logLevel: Log level to use for this instance.
  /// - parameter engine:   The underlying engine to use for starting Envoy.
  convenience init(config: EnvoyConfiguration, logLevel: LogLevel = .info, engine: EnvoyEngine) {
    self.init(configType: .typed(config), logLevel: logLevel, engine: engine)
  }

  /// Initialize a new Envoy instance using a string configuration.
  ///
  /// - parameter configYAML: Configuration yaml to use for starting Envoy.
  /// - parameter logLevel:   Log level to use for this instance.
  /// - parameter engine:     The underlying engine to use for starting Envoy.
  convenience init(configYAML: String, logLevel: LogLevel = .info, engine: EnvoyEngine) {
    self.init(configType: .yaml(configYAML), logLevel: logLevel, engine: engine)
  }
}

extension EnvoyClient: StreamClient {
  func newStreamPrototype() -> StreamPrototype {
    return StreamPrototype(engine: self.engine)
  }
}

File: mobile/library/kotlin/src/io/envoyproxy/envoymobile/EnvoyClient.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyConfiguration
import io.envoyproxy.envoymobile.engine.EnvoyEngine

/**
 * Envoy's implementation of `StreamClient`, buildable using `StreamClientBuilder`.
 */
internal class EnvoyClient private constructor(
  internal val engine: EnvoyEngine,
  internal val envoyConfiguration: EnvoyConfiguration?,
  internal val configurationYAML: String?,
  internal val logLevel: LogLevel
) : StreamClient {
  constructor(
    engine: EnvoyEngine,
    envoyConfiguration: EnvoyConfiguration,
    logLevel: LogLevel = LogLevel.INFO
  ) : this(engine, envoyConfiguration, null, logLevel)
  constructor(
    engine: EnvoyEngine,
    configurationYAML: String,
    logLevel: LogLevel = LogLevel.INFO
  ) : this(engine, null, configurationYAML, logLevel)

  init {
    if (envoyConfiguration == null) {
      engine.runWithConfig(configurationYAML, logLevel.level)
    } else {
      engine.runWithConfig(envoyConfiguration, logLevel.level)
    }
  }

  override fun newStreamPrototype() = StreamPrototype(engine)
}
